### DISCUSSION TITLE
Pre- and post processor for loading / storing function

### DISCUSSION THREAD

**Discussion by le-krogoth** [WerWolv/ImHex#649] [General]
I very much like what you did with ImHex. What a nice little piece of software.

While playing around with some specific file format, I missed a feature. And am not sure if I just wasn't able to find it, or if it is still missing. Let me explain: The binary file I am loading has the speciality, that the first few bytes can be read without any obfuscation. Then one byte (let's say the tenth) is the value to be used to xor the rest of the file.

Which means, it would be great if I could define for every field in the hexpat definition, if it should be preprocessed (and how, lets say the name of the function in the script) before showing in the hex editor. And we could then do the same again before saving the file. Running some kind of postprocessor just right before saving the file. With that feature I could manually manipulate the file within the hex editor without the need to run a third party script before opening said file.

Did I miss that feature in the doc, or have you planned such a feature already?

**Comment by WerWolv** [comment]
Hi
There is such a feature called the "Data Processor" which is a node-based scripting engine to pre-process your data. It's mostly intended for reading though since modifications made to the file with it will be stored in a so-called Overlay that's overlayed on top of your actual data. So things get displayed as if it was part of the file, however modifications that get made will still apply to the file itself (so at the end that byte will not be encrypted anymore)

**Reply by darkfiberiru** [reply]
Are there any data processor examples. I can't figure out how to get data to be overlayed. I can display with a buffer display but cant get it to overlay.

![image](https://user-images.githubusercontent.com/8406051/214337776-cb6aa03f-302a-49f5-b2f9-04b893cc29d8.png)

**Comment by le-krogoth** [comment]
All right, will have to look into this. Thanks for the hint!

Only problem I see from your description is that when I then store the manipulated file, the obfuscation is not applied back to the file. Which would be a problem.

**Reply by WerWolv** [reply]
Yeah. ImHex is primarily a binary analysis tool and just has some basic features to manipulate data. 
If you need to modify it, it would probably be better to have an external tool do it for you.

**Reply by le-krogoth** [reply]
I see. I am thinking of adding two hooks to the pattern editor script. One fired just after reading, the other just before saving. Do you think this could a) work and what would you b) think about difficulty implementing this? Taking from what I have seen so far, this is probably not working, since the pattern script is not necessarily run at a given moment.

**Reply by WerWolv** [reply]
I do have plans to use the pattern language for hooks / script like this for various other things too (like adding rows to the data inspector for example). I think those should be separate from the pattern editor though since the code in the pattern editor will just be executed once to generate the patterns and can't be auto-executed.

The language currently also has no way to write to your loaded data at all which would need to be implemented before as well.

So in short, I like the idea but it will need some work work and thought before it can be done nicely

**Reply by le-krogoth** [reply]
I agree. It would probably need a new kind of scripts. Which would then need a somewhat clear definition of what is used where and be triggered when. Including some thought on how could those new scripts interfere with the pattern language.

I'd probably go with reading the data, structure it as per pattern definition. Run the processing script, resulting in a new structure (potentially in a new pattern definition?). Showing that in the editor. And doing the same in reverse when saving. Which would allow for even wilder transformations, like migrating a file from version 1 of a file definition to a version 2.

Nevertheless. Great stuff, looking forward to those changes. If I can help out there, please shout.
